runtime/src/stake_history.rs

```diff
@@ -0,0 +1,84 @@
+//! This module implements clone-on-write semantics for the SDK's `StakeHistory` to reduce
+//! unnecessary cloning of the underlying vector.
+use std::{
+    ops::{Deref, DerefMut},
+    sync::Arc,
+};
+
+/// The SDK's stake history with clone-on-write semantics
+#[derive(Default, Clone, PartialEq, Debug, Deserialize, Serialize, AbiExample)]
+pub struct StakeHistory(Arc<StakeHistoryInner>);
```

Comment:
Similar to `StageDelegations` in #21542, we need an `Arc` here since banks are sent across threads (i.e. the accounts background services).